### DISCUSSION TITLE
Arkadyuti Bandyopadhyay - Github Externship Proposal

### DISCUSSION THREAD

**Discussion by ARKEOLOGIST** [hoppscotch/hoppscotch#1660] [Q&A]
**Application number:** 24-04_Ark377_tfa_258

**Full name:** Arkadyuti Bandyopadhyay

**E-mail:** h20200041@goa.bits-pilani.ac.in

**GitHub username:** ARKEOLOGIST

**About me:** I am currently pursuing an M.E. in Computer Science from Birla Institute of Technology and Science, Goa. Throughout the course of my B.Tech, which I had completed in 2020, I had completed three technical internships which got me a lot of exposure to the industry. The first application that I worked on was a VueJS-based application called Citizens of Rome: Dynasty Ascendant. It is a video game that is present on multiple platforms, including the [mobile](https://play.google.com/store/apps/details?id=net.rangergo.rome) as well as on [PC](https://store.steampowered.com/app/1063790/Citizen_of_Rome__Dynasty_Ascendant/). I put in some effort to learn some React and React Native in the last two years of my B.Tech, which allowed me to get two more internships. I developed one React Native application from scratch called [Taboo](https://play.google.com/store/apps/details?id=com.msocial.taboo) based on the design and functionality intended by my employers. I have also worked on another React native application called [Quintol](https://play.google.com/store/apps/details?id=com.quintol) where I had to code a few pages for the applications and change the appearance of a few existing pages. I have some minor NodeJS and vanilla React experience as well - as I have worked on small projects time and again to improve my experience.

More recently, I have become more regular with open source contributions. I have been a regular contributor to the [main web app](https://github.com/oppia/oppia) for the [Oppia Foundation](https://www.oppiafoundation.org/). I have become a member of [Oppia](https://github.com/oppia) more recently, becoming a part of the Learner's and Creator's team within the organization. 

**Prior experience:**
- [X] HTML
- [X] CSS
- [ ] SCSS
- [ ] Tailwind CSS
- [X] JavaScript
- [X] TypeScript
- [X] Vue.js
- [ ] Nuxt.js
- [ ] NestJS
- [ ] Firebase
- [ ] Prisma
- [ ] GraphQL
- [ ] PostgreSQL
- [ ] Go lang

---

**Proposal type:**
- [ ] Existing feature request
- [X] New feature proposal
- [ ] Refactor existing feature

**Proposal title:** Improving the dev flow for the Hoppscotch application

**Proposal description:** The main aim of this project is to improve the dev flow for the application. This will ensure that any future contributors working on it will have an easier experience working with it. This can be done in two phases:-

1) Improve the Jest-based tests for the application.

Hoppscotch uses Jest to set up individual tests for the application. The tests in their current state, are very basic by nature. This means that it might be easy to miss corner cases for the tests which might cause errors on certain launch states for the application. Improvements to the unit tests defined for the application as well as the addition of new, more comprehensive tests is the main goal of this phase. A secondary goal would definitely be to ensure that all of the modules are covered by tests at all times.

2) Document all modules for the application.

One of the main things that are possibly daunting for first-time contributors is the daunting project structure which is difficult for people to navigate. Adding documentation to help guide contributors wherever required is the main goal of this phase. The main aim should be to document all the modules of the project properly so that everyone working with the modules knows exactly what they do. Hoppscotch does have comments explaining some of the important modules - the goal would be to extend these comments to make the project structure as easy to comprehend as possible (the end goal should be that the repository structure should be comprehensible even to the first-time contributors). The recommendation is to use [JSDoc](https://github.com/jsdoc/jsdoc) for doing this.

**Proposal roadmap:** (brief list of things you will do/create, in some rough chronological order)
1. Analyze which portions of the project benefit from being covered by tests
2. Improve the existing unit tests
3. Write new tests wherever required
4. Identify key modules where the documentation needs to be improved
5. Add comments explaining critical functionality within the modules
6. Improve existing comments to be more 'verbose' 
7. Add or update documentation wherever required

**Comment by AndrewBastin** [comment]
Thank you for the proposal!

Please note that we do actually use Jest for testing in the repository. You can find the existing tests tucked inside a `__tests__` folder inside the relevant folder (have a peek at `helpers` and `components`). All our test files end with the extension `.spec.js` or `.spec.ts` (ts = TypeScript).

So, you might want to update the proposal to reflect this.

**Reply by ARKEOLOGIST** [reply]
Thanks for mentioning this, I'll update my proposal accordingly.

**Reply by AndrewBastin** [reply]
You can continue with this topic of unit tests though. We do have basic tests, but it's not really comprehensive. If you can bring to the table something more, that would be awesome.

**Reply by ARKEOLOGIST** [reply]
Sorry for the delayed response, but I have updated my proposal to make it as comprehensive as possible.
